plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/ArtifactPath.java

```diff
@@ -0,0 +1,298 @@
+ * Eclipse Foundation. All other trademarks are the property of their respective owners.
+ */
+package org.sonatype.nexus.repository.maven.internal;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.Immutable;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * Maven repository artifact path, a path that conforms Maven repository layout and can be "reverse engineered" into
+ * artifact GAV coordinates. Every artifact in repository have Maven GAV, may have hashes or
+ * signatures.
+ *
+ * @see Path
+ * @since 3.0
+ */
+@Immutable
+public class ArtifactPath
+    extends Path
+{
+  public static enum SignatureType
+  {
+    GPG("asc");
+
+    private final String ext;
+
+    SignatureType(final String ext) {
+      this.ext = ext;
+    }
+
+    public String getExt() {
+      return ext;
+    }
+  }
+
+  private final boolean snapshot;
+
+  private final String groupId;
+
+  private final String artifactId;
+
+  private final String version;
+
+  private final String baseVersion;
+
+  private final String classifier;
+
+  private final String extension;
+
+  private final SignatureType signatureType;
+
+  public ArtifactPath(final String path,
+                      final boolean snapshot,
+                      final String groupId,
+                      final String artifactId,
+                      final String version,
+                      final String baseVersion,
+                      final @Nullable String classifier,
+                      final String extension,
+                      final @Nullable SignatureType signatureType)
+  {
+    super(path);
+    this.snapshot = snapshot;
+    this.groupId = checkNotNull(groupId);
+    this.artifactId = checkNotNull(artifactId);
+    this.version = checkNotNull(version);
+    this.baseVersion = checkNotNull(baseVersion);
+    this.classifier = classifier;
+    this.extension = checkNotNull(extension);
+    this.signatureType = signatureType;
+  }
+
+  public boolean isSnapshot() {
+    return snapshot;
+  }
+
+  @Nonnull
+  public String getGroupId() {
+    return groupId;
+  }
+
+  @Nonnull
+  public String getArtifactId() {
+    return artifactId;
+  }
+
+  @Nonnull
+  public String getVersion() {
+    return version;
+  }
+
+  @Nonnull
+  public String getBaseVersion() {
+    return baseVersion;
+  }
+
+  @Nullable
+  public String getClassifier() {
+    return classifier;
+  }
+
+  @Nonnull
+  public String getExtension() {
+    return extension;
+  }
+
+  @Nullable
+  public SignatureType getSignatureType() {
+    return signatureType;
+  }
+
+  /**
+   * Returns {@code true} if this path is subordinate of another path.
+   *
+   * @see {@link #subordinateOf()}
+   */
+  @Override
+  public boolean isSubordinate() {
+    return isSignature() || super.isSubordinate();
+  }
```

Comment:
I question the value of this, why do we care if another path is subordinate or not?